core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/Envelope.java

```diff
@@ -117,6 +117,9 @@ public double getMinSpeed() {
         return minSpeed;
     }
 
+    /** Returns the envelope parts list */
+    public EnvelopePart[] getParts() { return Arrays.copyOf(parts, parts.length); }
+
```

Comment:
This copies the whole envelope, it's quite expensive. And not needed (see the other suggestion when iterating)